### DISCUSSION TITLE
Trigger a Command on frontend

### DISCUSSION THREAD

**Discussion by vibsid0986** [eclipse-langium/langium#1324] [Q&A]
Hello Team,

Below are my queries, request you to provide assistance on them.

1. I have created a worker thread on backend to facilitate the process of code generation for multiple files, however I am looking for a way to trigger a command on frontend once the generation is successful. 
2. Since there can be multiple files for which code generation will take place, is there a way to show progress bar on frontend.

Thanks in advance.

**Comment by msujew** [comment]
Hey @vibsid0986,

for both of these, you likely need the same mechanism. I assume your frontend is a language client, like vscode? In that case, it's just enough to send a notification to the language client from your language server that indicates that the generation has finished/is still ongoing.

```ts
// On generator finished/updated
connection.sendNotification('generator/progress');
```

Then on the client side, you can invoke the vscode API to display a notification/invoke a command:

```ts
languageClient.onNotification('generator/progress', () => {
  vscode.commands.executeCommand('<name-of-command>');
});
```

The progress report would work similar, but it would also send the progress value as an argument for the notification.

**Comment by vibsid0986** [comment]
Hi @msujew ,

Thanks for your quick response. It was really very helpful and I was able to implement it successfully, however in case of progress bar, seems like for every file processed it will invoke the command again and again, which would not be ideal. 

I am currently sending progress with the number of files processed as parameter through sendNotification from worker thread to main. Can you help me with a way to handle progress for such scenario.

**Reply by msujew** [reply]
You would probably have two commands on the language client. `startProgress` and `updateProgress`. They would work like this:

```ts
let progress: vscode.Progress<{
    message?: string | undefined;
    increment?: number | undefined;
}> | undefined;
let total = 0;
languageClient.onNotification('startProgress', (totalValue: number) => {
  total = totalValue;
  vscode.window.withProgress({
    location: vscode.ProgressLocation.Notification,
    title: 'Progress Report'
  }, async (progressValue) => {
    progress = progressValue;
  });
});
languageClient.onNotification('updateProgress', (increment: number) => {
  if (progress) {
    progress.report({ increment: (increment / total * 100) });
  }
});
```
